presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -214,6 +216,11 @@ else if (exchange.getType() == ExchangeNode.Type.REPLICATE) {
                     .map(PlanFragment::getId)
                     .collect(toImmutableList());
 
+            if (exchange.getType() == ExchangeNode.Type.MERGE_GATHER) {
+                checkState(exchange.getSources().get(0) instanceof SortNode, "Merge must be preceded by sort");
```

Comment:
Probably we just need to add an assertion here that there is always single source.